cmd/src/main_test.go

```diff
@@ -0,0 +1,179 @@
+			},
+			want: &config{
+				Endpoint:    "https://example.com",
+				AccessToken: "deadbeef",
+			},
+		},
+		{
+			name: "config file, token override only",
+			fileContents: &config{
+				Endpoint:    "https://example.com/",
+				AccessToken: "deadbeef",
+			},
+			envToken: "abc",
+			want:     nil,
+			wantErr:  errConfigMerge.Error(),
+		},
+		{
+			name: "config file, endpoint override only",
+			fileContents: &config{
+				Endpoint:    "https://example.com/",
+				AccessToken: "deadbeef",
+			},
+			envEndpoint: "https://exmaple2.com",
+			want:        nil,
+			wantErr:     errConfigMerge.Error(),
+		},
+		{
+			name: "config file, both override",
+			fileContents: &config{
+				Endpoint:    "https://example.com/",
+				AccessToken: "deadbeef",
+			},
+			envToken:    "abc",
+			envEndpoint: "https://override.com",
+			want: &config{
+				Endpoint:    "https://override.com",
+				AccessToken: "abc",
+			},
+		},
+		{
+			name:     "no config file, token from environment",
+			envToken: "abc",
+			want: &config{
+				Endpoint:    "https://sourcegraph.com",
+				AccessToken: "abc",
+			},
+		},
+		{
+			name:        "no config file, endpoint from environment",
+			envEndpoint: "https://example.com",
+			want: &config{
+				Endpoint:    "https://example.com",
+				AccessToken: "",
+			},
+		},
+		{
+			name:        "no config file, both variables",
+			envEndpoint: "https://example.com",
+			envToken:    "abc",
+			want: &config{
+				Endpoint:    "https://example.com",
+				AccessToken: "abc",
+			},
+		},
+		{
+			name:         "endpoint flag should override config",
+			flagEndpoint: "https://override.com/",
+			fileContents: &config{
+				Endpoint:    "https://example.com/",
+				AccessToken: "deadbeef",
+			},
+			want: &config{
+				Endpoint:    "https://override.com",
+				AccessToken: "deadbeef",
+			},
+		},
+		{
+			name:         "endpoint flag should override environment",
+			flagEndpoint: "https://override.com/",
+			envEndpoint:  "https://example.com",
+			envToken:     "abc",
+			want: &config{
+				Endpoint:    "https://override.com",
+				AccessToken: "abc",
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			oldConfigPath := *configPath
+			defer func() { *configPath = oldConfigPath }()
+
+			if test.flagEndpoint != "" {
+				val := test.flagEndpoint
+				endpoint = &val
+				defer func() { endpoint = nil }()
+			}
+
+			if test.fileContents != nil {
+				p, cleanup := makeTempConfig(t, *test.fileContents)
+				defer cleanup()
+				*configPath = p
+			}
+			oldToken := os.Getenv("SRC_ACCESS_TOKEN")
+			defer func() { os.Setenv("SRC_ACCESS_TOKEN", oldToken) }()
+			oldEndpoint := os.Getenv("SRC_ENDPOINT")
+			defer func() { os.Setenv("SRC_ENDPOINT", oldEndpoint) }()
+
+			if err := os.Setenv("SRC_ACCESS_TOKEN", test.envToken); err != nil {
+				t.Fatal(err)
+			}
+			if err := os.Setenv("SRC_ENDPOINT", test.envEndpoint); err != nil {
+				t.Fatal(err)
+			}
```

Comment:
Maybe a bit easier to parse visually and read 😄 But, I was in the mood to write a tiny function just now, feel free to ignore:

```suggestion
			setEnvVar := func(name, val string) {
				old := os.Getenv(name)
				if err := os.Setenv(name, val); err != nil {
					t.Fatal(err)
				}
				t.Cleanup(func() { os.Setenv(name, old) })
			}
			setEnvVar("SRC_ACCESS_TOKEN", test.envToken)
			setEnvVar("SRC_ENDPOINT", test.envEndpoint)
```